File: test/corpus/propagation/list_element_propagation.exs
defmodule RustQ.Corpus.Propagation.ListElementPropagation do
  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec decode(R.term()) :: R.nif_result(R.path(:Color))
  defrust decode(term) do
    value = decode_as!(term, R.u32())
    {:ok, Color.from_argb(255, 0, 0, value)}
  end

  @spec consume(R.vec(R.path(:Color))) :: R.nif_result(R.unit())
  defrust(consume(_colors), do: :ok)

  @spec run(R.term()) :: R.nif_result(R.unit())
  defrust run(term) do
    consume([decode(term)])
    :ok
  end
end

File: test/corpus/propagation/list_element_propagation.rs
fn decode<'a>(term: Term<'a>) -> NifResult<Color> {
    let value = term.decode::<u32>()?;
    Ok(Color::from_argb(255, 0, 0, value))
}

fn consume(_colors: Vec<Color>) -> NifResult<()> {
    Ok(())
}

fn run<'a>(term: Term<'a>) -> NifResult<()> {
    consume(vec![decode(term)?])?;
    Ok(())
}
